src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -72,6 +77,189 @@ internal class Http2FrameWriter
         _scheduleInline = serviceContext.Scheduler == PipeScheduler.Inline;
 
         _hpackEncoder = new DynamicHPackEncoder(serviceContext.ServerOptions.AllowResponseHeaderCompression);
+
+        // In practice, this is bounded by the number of concurrent streams allowed + whatever overflow we allow
+        _channel = Channel.CreateUnbounded<Http2OutputProducer>(new UnboundedChannelOptions()
+        {
+            AllowSynchronousContinuations = _scheduleInline,
+            SingleReader = true
+        });
+
+        _window = initialWindowSize;
+
+        _writeQueueTask = Task.Run(WriteToOutputPipe);
+    }
+
+    public void Schedule(Http2OutputProducer producer)
+    {
+        _channel.Writer.TryWrite(producer);
+    }
+
+    private async Task WriteToOutputPipe()
+    {
+        await foreach (var producer in _channel.Reader.ReadAllAsync())
+        {
+            try
+            {
+                var reader = producer.PipeReader;
+                var stream = producer.Stream;
+
+                var observed = Http2OutputProducer.State.None;
+
+                // We don't need to check the result because it's either
+                // - true because we have a result
+                // - false because we're flushing headers
+                var hasResult = reader.TryRead(out var readResult);
```

Comment:
nit

Why not `reader.TryRead(out var readResult);` with no hasResult variable? Or is it a warning to not use the result?

Alternatively, `_ = reader.TryRead(out var readResult);` to discard.

Reply:
oops. Mistake.